[WIP] Synthesized instruments

DeltaEpsilon7787 opened the pull request:
Mentioned a few months ago in #17601 

This is a framework for adding new musical instruments to the game. It allows you to get a full spread out of only a few samples. It also allows you to control sound precisely. This should make musician.dm and similar completely obsolete.
I made sure it's capable of playing as smoothly as possible.
You can see it in action if you spawn /obj/structure/synthesized_instruments/synthesizer.

It includes one universal instrument "Synthesizer 3.0" that allows you to use all features present in this system. 

This system is completely modular and its only dependency is NanoUI. You can easily add and remove it from game.

Don't merge now

Things to do:
- [ ] (IN PROGRESS) Remove redundant and old parts of code
- [ ] (IN PROGRESS) Write comments
- [ ] (IN PROGRESS) Separate internal and external parts of this system
- [x] (HIGH PRIORITY) Fix Song Editor NanoUI template
- [x] /obj/sound_player should become /datum/sound_player because there is no reason to have physical presence of /obj/sound_player
- [x] There is no point in having /datum/synthesized_song. Move all of its functionality to /datum/sound_player and delete it
- [ ] Write docs
- [x] Decide on what to do with music code
- [x] Fix Destroy()
- [x] Prune debug away
- [ ] Integrate playsound_local functionality into this system
- [ ] Convert regular piano, violin and guitar into this system
- [ ] Remove musician.dm and stuff
- [ ] Place the torture synthesizer in the torture chamber (aka interrogation)
- [x] Get permission from soundfonts' authors to use their samples (all of used soundfonts are free and available on the Internet for download, but I'm not sure if I'm allowed to use their sounds here)
- [ ] Re-record sound samples and normalize them.
- [x] Stop using BYOND's GC.
- [x] global.musical_config should be moved to /decl/global_vars
- [x] Make a proper module out of it (that is, make "code", "sound", "icon" folders and stuff and make it a module inside a module)
- [x] Remove duplicate code in real_instruments.dm by using #include preprocessor directive
- [ ] (LOW PRIORITY) This sprite is shit, replace it
- [ ] (LOW PRIORITY) Add new instruments

DeltaEpsilon7787 committed:
Synthesized Instruments (version 3.0)

DeltaEpsilon7787 committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into dev

CrimsonShrike commented:
I see file count is quite high. Is it because of number of samples and such?

DeltaEpsilon7787 commented:
Yes. There are only about 8 code files and 5 templates.
Everything else is ogg files.
Be aware that 80% of them are actually copies from sound/piano and sound/violin

DeltaEpsilon7787 committed:
/obj/sound_player --> /datum/sound_player
global namespace --> GLOB namespace
Code structure is now more clear
IT COMPILES
Destroy() should work properly now

DeltaEpsilon7787 committed:
Song Editor unfucked

DeltaEpsilon7787 commented:
Well shit. Can't use #include. Guess I'll go with copypaste after all.

DeltaEpsilon7787 committed:
Let's try to pass code quality checks (I hate you all)

DeltaEpsilon7787 committed:
Fuck

DeltaEpsilon7787 commented:
It works now, but song editor is still broken
![Like this](https://vgy.me/TTkNND.png)

DeltaEpsilon7787 commented:
Not anymore

PsiOmegaDelta commented:
We do already have a sound player: https://github.com/Baystation12/Baystation12/blob/dev/code/datums/sound_player.dm albeit for looping sounds.

DeltaEpsilon7787 commented:
1. It would add a dependency and I'm trying to avoid this at all costs
2. This module didn't exist back when I started writing synthesized instruments. Rewriting code to use it will be a huge pain. I'm also going to pull request synthesized instrument to TG when I'm done here and they do not have this module.
3. Also, my version of it has some functionality that datum/sound_player.dm doesn't.
For example, applying sound.echo and sound.environment and controlling whether a sound is 3D or not.

DeltaEpsilon7787 committed:
Debug no more

PsiOmegaDelta commented:
When I end up reviewing this in earnest and point of sections of code that would be improved by utilizing existing systems I will expect them to be used.
Having multiple sections of near-identical code simply because you don't want dependencies won't be acceptable.

We'll see on which side the sound player, and everything else ends up.

DeltaEpsilon7787 committed:
It never existed in the first place

DeltaEpsilon7787 commented:
Actually, now that I think of it /datum/sound_player, as of now, is only used to apply initial modifications to sound variables and also finding people to play to. It **is not** used to control sound. This is actually done in /datum/synthesized_song and /datum/musical_event_manager. Well, I'm going to remove /datum/synthesized_song anyway, so it will be doing the job of both controlling sound AND parsing music.

It is actually event manager that is applying dynamic settings to sounds (namely changing volume). Static settings like echo or x,y,z are done by datum/sound_player.

Sorry for /datum/synthesized_song, this system is based on musician.dm after all.

DeltaEpsilon7787 committed:
Y u count << in /* << << << */ shit dammit

DeltaEpsilon7787 commented:
@PsiOmegaDelta 
Wait a minute.
/decl/sound_player is for _looping_ sounds. Like ambient sounds from machinery. This isn't something you'd want from a musical instrument.
I cannot use /decl/sound_player unless someone adds proper functionality to it.

DeltaEpsilon7787 committed:
Gah, DME files

ghost commented:
Hell yes

DeltaEpsilon7787 commented:
Oh come on, I can't even use << operator for bit shifts?

PsiOmegaDelta commented:
You can, but you'll have to manually up the Travis limit.
However, seems it'd just be easier to use constants, either numerals 0x001, 0x002, ..., 0x0080, or named values (SOUND_IS_PLAYING, SOUND_PAUSED, or w.e. is relevant for your particular use cases).

DeltaEpsilon7787 commented:
It's fine, all bit-shifts are done in debug.dm in checks just to store which checks failed and which did not. This is going to be removed as well.

DeltaEpsilon7787 commented:
@PsiOmegaDelta 
Why does it also count <<'s in commented code?

chaoko99 commented:
If you could move this onto, say, Midi in some way, you could save a crapload of file additions... And a lot of reading, due to the fact that midis are generally handled by one's sound card.

Also: They're less likely to cause a noticable stalling on the server

DeltaEpsilon7787 commented:
@chaoko99, you don't seem to understand how and why midi works.

Here all the samples are basically prerecorded sounds that get played at a faster or slower rate (and as such, different pitch), which creates the whole spread.

MIDI contains no sound information, it's simply an instruction. What and how sounds are played is defined in soundfonts.

chaoko99 commented:
Which is precisely why I was suggesting you use midi. Our current music system stalls the entire server between notes. It functions in a similar (If not identical) manner to what this does.

Midi is moved onto the sound card, playing off the sound fonts... A plaintext instruction is a lot faster than a whole file.

AtomicWorm commented:
I won't pretend I understand any of this thing, but if it allows me to take a piano sheet and transcribe it into something that will actually sound like a song in-game, you're my new hero dude.

DeltaEpsilon7787 commented:
@chaoko99 I think you are wrong.
What server does is basically send a command to a client to play a certain sound file from their **local cache.** There is no downloading of sounds or anything like that.

Additionally, I've done plenty of profiling of my system and the slowest procedure is not even what plays the sounds, but what finds listeners, which I fixed. (it would use O(n^2) func (range(15. src) - view(15, src)), now it iterates through all mobs instead which is O(n))

@AtomicWorm 
This thing has been around for years, I've even made a video of the very first version of the synthesizer:
https://www.youtube.com/watch?v=exQDY0eohHA

Even then it sounded much better than current piano.

DeltaEpsilon7787 committed:
/datum/namespace/synthesized_instruments is a thing now and many more

All functionality was moved into a namespace, also started working on docstrings and separating internal and external functions and variables

Started working on ADSR cycle instead of simple sustain timer.

Refactored some code

BlueNexus commented on code/modules/synthesized_instruments/code/globals.dm:
```diff
@@ -0,0 +1,339 @@
+	return src.num2text_cached[key]
+
+Even though it is 3 times as fast as stock num2text, it only gives marginable performance boost
+Dumbing down this section.
```
We don't accept commented-out code. Remove this.

DeltaEpsilon7787 replied:
Sure

BlueNexus commented on code/modules/synthesized_instruments/code/core/instruments.dm:
```diff
@@ -0,0 +1,56 @@
+	var/last_sample = src.samples[GLOB.musical_config.n2t(last_key)]
+	for (var/key=0 to first_key-1) src.sample_map[GLOB.musical_config.n2t(key)] = new /datum/sample_pair(first_sample, key-first_key)
+	for (var/key=last_key to 127)	src.sample_map[GLOB.musical_config.n2t(key)] = new /datum/sample_pair(last_sample,  key-last_key)
+	return src.samples
```
Ditto

BlueNexus reviewed (changes requested):
I'm aware this is a WIP, but there's a few things you need to do at some point.

mustafakalash commented:
It has been almost a month, are you still working on this?

DeltaEpsilon7787 commented:
Yes, but I'm really busy with real life stuff now.

@mustafakalash

DeltaEpsilon7787 commented:
I'm back at it again, finally.

DeltaEpsilon7787 committed:
Just to show people that this ain't dead yet

comma commented:
Going to call it
